### PR TITLE
Dynamically fetch FHIR definitions from fhir-api

### PR DESCRIPTION
**Pull request by simonvadee**
server:
- `bootstrapDefinition` uses static Bundle files instead of fetching definitions from the API
- remove `profiles`and `extensions` resolvers on StructureDefinition
- `getDefinition` sends a query to fhir-api if the deifnition was not cached yet

client:
- cache availableResources and availableExtensions in a redux reducer `fhir` instead of fetching them from pyrog-api
- resource profiles are loaded dynamically

### CODE REVIEW

File: client/src/components/mapping/FhirMappingPanel/index.tsx [hunk 1]
```diff
@@ -48,6 +49,10 @@ const FhirMappingPanel = () => {
     );
   };
 
+  React.useEffect(() =>{
+    dispatch(fetchAvailableExtensions)
```
Comment (Jasopaum) [inline client/src/components/mapping/FhirMappingPanel/index.tsx:53]: Is this called when a new extension is created?

File: client/src/components/selects/addResourceSelect.tsx [hunk 2]
```diff
@@ -1,46 +1,32 @@
...
-  profiles: Profile[];
-}
-
-const filterByName: ItemPredicate<Profile> = (query, resource: Profile) => {
-  return resource.name.toLowerCase().indexOf(query.toLowerCase()) >= 0;
+const filterByName: ItemPredicate<ResourceDefinition> = (query, definition: ResourceDefinition) => {
+  return definition.name.toLowerCase().indexOf(query.toLowerCase()) >= 0;
 };
 
 const sortAlphabetically = (
```
Comment (Jasopaum) [inline client/src/components/selects/addResourceSelect.tsx:27]: We could refactor this using `localeCompare`, I think Teva did that in some other components

File: client/src/components/selects/addProfileSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+import TSelect from './TSelect';
+
+interface Props {
+  onChange: any;
+  definition: ResourceDefinition;
+  openUploadProfile: (r: ResourceDefinition) => void;
+  fetchProfiles: (r: ResourceDefinition) => Promise<ResourceDefinition[]>
+}
+
+const sortAlphabetically = (
```
Comment (Jasopaum) [inline client/src/components/selects/addProfileSelect.tsx:24]: Same, we could use `localeCompare`

File: client/src/components/mapping/FhirMappingPanel/ResourceSelector/Drawer/index.tsx [hunk 4]
```diff
@@ -52,7 +46,7 @@ const Drawer = ({ isOpen, onCloseCallback }: Props): ReactElement => {
   const [label, setLabel] = React.useState('');
   const [pkTable, setPkTable] = React.useState('');
   const [pkColumn, setPkColumn] = React.useState('');
-  const [filters, setFilters] = React.useState([] as Filter[]);
+  const [filters, setFilters] = React.useState([] as Filters[]);
```
Comment (Jasopaum) [inline client/src/components/mapping/FhirMappingPanel/ResourceSelector/Drawer/index.tsx:49]: nit + I know it's not you: could we rename `Filters` -> `Filter`?

File: client/src/services/fhir/index.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        return acc;
+      }, {});
+    dispatch({ type: 'FETCH_EXTENSIONS', payload: extensions });
+  } catch (err) {
+    console.error(
+      `Could not fetch available extensions: ${
+        err.response ? err.response.data : err.message
+      }`
+    );
+    dispatch({ type: 'FETCH_EXTENSIONS', payload: getState() });
```
Comment (Jasopaum) [inline client/src/services/fhir/index.ts:38]: Shouldn't it be `getState().extensions`?

File: client/src/services/fhir/index.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    });
+
+    return;
+  } catch (err) {
+    console.error(
+      `Could not fetch available resources: ${
+        err.response ? err.response.data : err.message
+      }`
+    );
+    dispatch({ type: 'FETCH_AVAILABLE_RESOURCES', payload: getState() });
```
Comment (Jasopaum) [inline client/src/services/fhir/index.ts:65]: Shouldn't it be `getState().availableResources`?

File: client/src/services/fhir/index.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  }
+};
+
+const extensionsReducer = (
```
Comment (Jasopaum) [inline client/src/services/fhir/index.ts:69]: `extensionsReducer` -> `fhirReducer`?

File: client/src/components/mapping/FhirMappingPanel/FhirResourceTree/index.tsx [hunk 8]
```diff
@@ -153,15 +160,25 @@ const FhirResourceTree = ({ onClickCallback }: Props) => {
...
 
+    const { data, errors } = await client.query({
+      query: qStructureDisplay,
+      variables: { definitionId: extensionDefinition.id }
+    });
+    if (!data) {
+      console.error('could not add extension:', errors)
+      return
+    }
+    console.log("yay", data.structureDefinition)
```
Comment (Jasopaum) [inline client/src/components/mapping/FhirMappingPanel/FhirResourceTree/index.tsx:171]: 🙊

File: client/src/components/mapping/FhirMappingPanel/FhirResourceTree/index.tsx [hunk 9]
```diff
@@ -153,15 +160,25 @@ const FhirResourceTree = ({ onClickCallback }: Props) => {
...
-      extensionDefinition
+      data.structureDefinition
     );
 
     node.isExpanded = true;
     extensionArrayNode.isExpanded = true;
     extensionNode!.isExpanded = true;
     if (selectedNode) selectedNode.isSelected = false;
     extensionNode!.isSelected = true;
+    console.log("setSelectedNode", extensionNode)
```
Comment (Jasopaum) [inline client/src/components/mapping/FhirMappingPanel/FhirResourceTree/index.tsx:181]: 🙈

File: client/src/services/fhir/index.ts (hunk 6 above)
Comment (simonvadee) [reply client/src/services/fhir/index.ts:65]: yes! done

File: client/src/services/fhir/index.ts (hunk 5 above)
Comment (simonvadee) [reply client/src/services/fhir/index.ts:38]: yes! done

File: client/src/services/fhir/index.ts (hunk 7 above)
Comment (simonvadee) [reply client/src/services/fhir/index.ts:69]: right, done

File: client/src/components/mapping/FhirMappingPanel/index.tsx (hunk 1 above)
Comment (simonvadee) [reply client/src/components/mapping/FhirMappingPanel/index.tsx:53]: it is now 👍

File: client/src/components/selects/addResourceSelect.tsx (hunk 2 above)
Comment (simonvadee) [reply client/src/components/selects/addResourceSelect.tsx:27]: done

File: client/src/components/selects/addProfileSelect.tsx (hunk 3 above)
Comment (simonvadee) [reply client/src/components/selects/addProfileSelect.tsx:24]: done

File: client/src/components/mapping/FhirMappingPanel/ResourceSelector/Drawer/index.tsx (hunk 4 above)
Comment (simonvadee) [reply client/src/components/mapping/FhirMappingPanel/ResourceSelector/Drawer/index.tsx:49]: done

File: client/src/components/selects/addResourceSelect.tsx [hunk 10]
```diff
@@ -24,28 +24,10 @@ const filterByName: ItemPredicate<ResourceDefinition> = (query, definition: Reso
...
-
-// If the second argument is provided, the sort funcion puts
-// the definition which id matches on top of the list.
-const sortItems = (
-  definitions: ResourceDefinition[],
-  first: string | undefined = undefined
-): ResourceDefinition[] => definitions.sort((r1, r2) => sortAlphabetically(r1, r2, first));
+const sortItems = (resources: ResourceDefinition[]): ResourceDefinition[] =>
+resources.sort((r1: ResourceDefinition, r2: ResourceDefinition): number =>
+  r1.name.localeCompare(r2.name)
```
Comment (Jasopaum) [inline client/src/components/selects/addResourceSelect.tsx:29]: Did you remove the `first` arg purposefully?

File: client/src/components/mapping/FhirMappingPanel/FhirResourceTree/nodeLabel.tsx [hunk 11]
```diff
@@ -79,15 +81,17 @@ export const NodeLabel = ({
...
-      <ApolloProvider client={client}>
-        <Menu>
-          {isArray && !isExtension && slices.length === 0 && renderAddItem()}
-          {isArray && slices.length > 0 && renderAddSlice()}
-          {isItem && renderRemoveItem()}
-          {hasAllowedExtensions && renderAddExtension(extensions)}
-          {isRootExtensions && renderAddExtension(resourceExtensions)}
-        </Menu>
-      </ApolloProvider>
+      <Provider store={store}>
```
Comment (Jasopaum) [inline client/src/components/mapping/FhirMappingPanel/FhirResourceTree/nodeLabel.tsx:84]: Why do we need that now? 🤔

**Review by Jasopaum — APPROVED** [review]

**Pull request merged**